raft-tech/arcadedb
# 1370 database and attribute classification

## What does this PR do?
1. Adds node/edge filtering based on classification and user clearance
2. Adds support for source based classification marking of nodes/edges
3. Adds initial support for ACCM checks, specifically NOFORN and RELEASEABLE TO based on nationality
4. Extends create database screen in Arcade to take classification, owner, public/private properties. These properties persist, and are viewable on the database page
5. Updated database dropdown in arcade to include portion classification markings
6. Added a checkbox to the create database screen in arcade to load Palantir's initial ontology based on wikidata
7. Prevent the users from classification marking objects higher than the classification of the database
8. Added a check to prevent users from creating databases (and objects) at a higher classification than the Df deployment

## Motivation
Customer requirement to secure data access

## Related issues
https://github.com/raft-tech/data-fabric/issues/1370
https://github.com/raft-tech/data-fabric/issues/2345

## Additional Notes
Requires lots of keycloak admin console tweaks

## Checklist
1. Build arcade locally

```
 docker buildx build --platform linux/arm64 --load -t raft-tech/arcadedev ./
 ```

2. Load into kind

```
 kind load docker-image raft-tech/arcadedev  --name data-fabric
```

3. Update image in arcade's stateful set

5. Kill the pod to force it to update if needed

6. In keycloak, remove the Data Steward group from the admin user.

7. Create a new `clearance-USA` attribute, with an initial value of `U`.

8. Create a new `nationality` attribute, and set it to `USA`.

## Test steps
1. log into arcade http://arcadedb.localhost/ with admin credentials
2. Create new database in arcade gui.(Click on the database icon, 2nd from the top on the left menu). Create button is in the upper right.
Set classification to S
Leave import ontology unchecked.

3. Switch to new database if needed in the dropdown
4. Switch back to the top button where you can run scripts.
5.  Create `People` vertex type
```
CREATE VERTEX TYPE People
```

6. Confirm the following fails because no clearance info is on the document

```
INSERT INTO People CONTENT 
{ 
  firstName: 'Enzo', 
  lastName: 'Ferrari'
}
```

7. Confirm the following command fails because TS is higher than the classification of the database (S)

```
INSERT INTO People CONTENT 
{ 
  firstName: 'Enzo', 
  lastName: 'Ferrari', 
  classification: { general: 'TS'}
}
```

8. Confirm the following fails

```
INSERT INTO People CONTENT 
{ 
  firstName: 'Enzo', 
  lastName: 'Ferrari', 
  classification: { general: 'S'}
}
```
9. Confirm the following fails

```
INSERT INTO People CONTENT 
{ 
  firstName: 'Enzo U',
  lastName: 'Ferrari',
  sources: {
    '1': '(S) ABC 33322'
  }
}
```

10. Confirm the following succeeds

```
INSERT INTO People CONTENT 
{ 
  firstName: 'Enzo1', 
  lastName: 'Ferrari', 
  classification: { general: 'U'}
}
```
11. Confirm the following succeeds

```
INSERT INTO People CONTENT 
{ 
  firstName: 'Enzo2',
  lastName: 'Ferrari',
  sources: {
    '1': '(U) ABC 33322'
  }
}
```

12. In keycloak, bump up the clearance attribute to S

13. Confirm the following now succeeds

 ```
INSERT INTO People CONTENT 
{ 
  firstName: 'Enzo3', 
  lastName: 'Ferrari', 
  classification: { general: 'S'}
}
```

14. Confirm the following now succeeds

```
INSERT INTO People CONTENT 
{ 
  firstName: 'Enzo4',
  lastName: 'Ferrari',
  sources: {
    '1': '(S) ABC 33322'
  }
}
```

15. Confirm the following succeeds

```
INSERT INTO People CONTENT 
{ 
  firstName: 'Enzo5',
  lastName: 'Ferrari',
  sources: {
    '1': '(S//NF) IIR 12345'
  }
}
```

16. Run the following command, confirm you can see all 5 created documents.

```
select from `People` limit 30
```

17. In keycloak, change the nationality attribute from USA to GBR.

18. In arcade, confirm the following command fails

```
INSERT INTO People CONTENT 
{ 
  firstName: 'Enzo6',
  lastName: 'Ferrari',
  sources: {
    '1': '(S//NF) IIR 12345'
  }
}
```

19. Run the following command, confirm you can see only see 4 created documents. Confirm the previously created Enzo5 document with a classification marking of `S//NF` isn't present

```
select from `People` limit 30
```

20. In keycloak, update the clearance attribute to U.

21. Run the following query, confirm only unclass docs Enzo1 and Enzo2 return

```
select from `People` limit 30
```

22. In keycloak, revert the value of the clearance-USA attribute back to `S`

23. In keycloak, revert the nationality attribute back to `USA`

24. Now we are going to validate edges loading only for authorized users.

25. Create a new Location vertex type

```
CREATE VERTEX TYPE Location
```

26. Create a new Location document

```
INSERT INTO Location CONTENT 
{ 
  name: 'Richmond',
  type: 'City',
  classification: {
    general: 'U'
  }
}
```
27. Create a new edge type

```
CREATE EDGE TYPE LivesIn
```

28. Confirm edge creation fails due to missing classification info

```
CREATE EDGE LivesIn 
FROM (SELECT FROM People WHERE firstName = 'Enzo1') TO
     (SELECT FROM Location WHERE name = 'Richmond')
```

29. confirm this succeeds

```
CREATE EDGE LivesIn 
FROM (SELECT FROM People WHERE firstName = 'Enzo1') TO
     (SELECT FROM Location WHERE name = 'Richmond')
CONTENT {
  sources: {
    '1': '(S//NF) IIR 12345'
  }
}
```

30. Run the following select to load the person for which we just added an edge.

```
SELECT FROM People WHERE firstName = 'Enzo1'
```

31. On the right side of the graph result window, click the both button to load all adjacent edges.

32. confirm you can see the LivesIn edge and Location Object in results. May be helpful to look at the graph result display.

33. in keycloak, revert the classification attribute back to U

34. In arcade, click the load both adjacent edges button again, confirm only the person loads.
